src/Stream/StreamWriter.php

```diff
@@ -0,0 +1,157 @@
+    /** @var string */
+    private $onError;
+    /** @var float|null */
+    private $timeout;
+    /** @var string */
+    private $instance;
+    /** @var int */
+    private $numFails = 0;
+    /** @var float */
+    private $waitTill = 0.0;
+
+    /**
+     * @param string     $instance
+     * @param string     $host
+     * @param int        $port
+     * @param string     $onError What to do on connection error
+     * @param float|null $timeout
+     */
+    public function __construct(
+        $instance = 'writer',
+        $host = '127.0.0.1',
+        $port = 8125,
+        $onError = self::ON_ERROR_EXCEPTION,
+        $timeout = null
+    ) {
+        $this->instance = $instance;
+        $this->host = $host;
+        $this->port = $port;
+        $this->onError = $onError;
+        $this->timeout = $timeout;
+    }
+
+    public function __destruct()
+    {
+        if ($this->socket) {
+            // the reason for this failing is that it is already closed, so ignore the result and not messing with
+            // parent classes
+            @fclose($this->socket);
+        }
+    }
+
+    /**
+     * @param string $message
+     *
+     * @return bool
+     */
+    public function write($message)
+    {
+        $this->ensureConnection();
+        if ($this->socket) {
+            if (@fwrite($this->socket, $message) === false) {
+                // attempt to re-send on socket resource failure
+                $this->socket = $this->connect();
+                if ($this->socket) {
+                    return (@fwrite($this->socket, $message) !== false);
+                }
+            } else {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Ensure that we are currently connected to the socket
+     */
+    protected function ensureConnection()
+    {
+        if ((!$this->socket) && ($this->canConnect())) {
+            $this->socket = $this->connect();
+        }
+    }
+
+    /**
+     * @return bool
+     */
+    protected function canConnect()
+    {
+        return (microtime(true) > $this->waitTill);
+    }
+
+    /**
+     * Attempt to connect to a stream
+     *
+     * @return null|resource
+     */
+    protected function connect()
+    {
+        $socket = @fsockopen('udp://' . $this->host, $this->port, $errno, $errstr, $this->timeout);
+        if ($socket === false) {
+            $this->waitTill = microtime(true) + (static::RETRY_INTERVAL * (pow(2, $this->numFails++)));
+
+            switch ($this->onError) {
+                case static::ON_ERROR_ERROR:
+                    trigger_error(
+                        sprintf('StatsD server connection failed (udp://%s:%d)', $this->host, $this->port),
+                        E_USER_WARNING
+                    );
+                    break;
+                case static::ON_ERROR_EXCEPTION:
+                    throw new ConnectionException($this->instance, '(' . $errno . ') ' . $errstr);
+            }
```

Comment:
Yeah, which i think is fine. You have to say I want to ignore it, and it has an exponential back-off to won't kill your application.